HPCC-18041 Unicode Implementation for ExcludeLastWord

dskaff opened the pull request:
- Adds testcases to ecllibarry/teststd/uni/TestExcludeLastWord.ecl
- Adds documentation to ecllibrary/std/Uni.ecl
- Adds code to plugins/unicodelib.cpp

Signed-off-by: David Skaff <david.skaff@lexisnexisrisk.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [x] My change requires a change to the documentation.
  - [x] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [x] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

HPCCSmoketest commented:
Automated Smoketest: :x:
Sha: 5bb8673846c237f1147c85194ff800a9a2cfc02f
Build: failed
Number of Error(s): 14
[  0%] Building CXX object aws-cpp-sdk-core/CMakeFiles/aws-cpp-sdk-core.dir/source/client/CoreErrors.cpp.o
/mnt/disk1/home/vamosax/smoketest/PR-10321/HPCC-Platform/plugins/unicodelib/unicodelib.hpp:104:1: error: expected unqualified-id before \xe2\x80\x98<<\xe2\x80\x99 token
/mnt/disk1/home/vamosax/smoketest/PR-10321/HPCC-Platform/plugins/unicodelib/unicodelib.hpp:106:1: error: expected unqualified-id before \xe2\x80\x98==\xe2\x80\x99 token
/mnt/disk1/home/vamosax/smoketest/PR-10321/HPCC-Platform/plugins/unicodelib/unicodelib.hpp:108:1: error: expected unqualified-id before \xe2\x80\x98>>\xe2\x80\x99 token
/mnt/disk1/home/vamosax/smoketest/PR-10321/HPCC-Platform/plugins/unicodelib/unicodelib.cpp:85:3: error: expected primary-expression before \xe2\x80\x98<<\xe2\x80\x99 token
/mnt/disk1/home/vamosax/smoketest/PR-10321/HPCC-Platform/plugins/unicodelib/unicodelib.cpp:85:5: error: expected primary-expression before \xe2\x80\x98<<\xe2\x80\x99 token
/mnt/disk1/home/vamosax/smoketest/PR-10321/HPCC-Platform/plugins/unicodelib/unicodelib.cpp:85:7: error: expected primary-expression before \xe2\x80\x98<\xe2\x80\x99 token
/mnt/disk1/home/vamosax/smoketest/PR-10321/HPCC-Platform/plugins/unicodelib/unicodelib.cpp:85:9: error: \xe2\x80\x98HEAD\xe2\x80\x99 was not declared in this scope
/mnt/disk1/home/vamosax/smoketest/PR-10321/HPCC-Platform/plugins/unicodelib/unicodelib.cpp:726:1: error: expected unqualified-id before \xe2\x80\x98<<\xe2\x80\x99 token
/mnt/disk1/home/vamosax/smoketest/PR-10321/HPCC-Platform/plugins/unicodelib/unicodelib.cpp:1522:1: error: expected \xe2\x80\x98}\xe2\x80\x99 at end of input
make[2]: *** [plugins/unicodelib/CMakeFiles/unicodelib.dir/unicodelib.cpp.o] Error 1
make[1]: *** [plugins/unicodelib/CMakeFiles/unicodelib.dir/all] Error 2
[100%] Building CXX object aws-cpp-sdk-sqs/CMakeFiles/aws-cpp-sdk-sqs.dir/source/SQSErrors.cpp.o
make: *** [all] Error 2
Get ECLWatch build result

HPCCSmoketest commented:
Automated Smoketest: :x:
Sha: a2e39252c92d2712881dc6fc397d0fd4e06e2dcc
Build: failed
Number of Error(s): 14
/mnt/disk1/home/vamosax/smoketest/PR-10321/HPCC-Platform/plugins/unicodelib/unicodelib.hpp:104:1: error: expected unqualified-id before \xe2\x80\x98<<\xe2\x80\x99 token
/mnt/disk1/home/vamosax/smoketest/PR-10321/HPCC-Platform/plugins/unicodelib/unicodelib.hpp:106:1: error: expected unqualified-id before \xe2\x80\x98==\xe2\x80\x99 token
/mnt/disk1/home/vamosax/smoketest/PR-10321/HPCC-Platform/plugins/unicodelib/unicodelib.hpp:108:1: error: expected unqualified-id before \xe2\x80\x98>>\xe2\x80\x99 token
/mnt/disk1/home/vamosax/smoketest/PR-10321/HPCC-Platform/plugins/unicodelib/unicodelib.cpp:85:3: error: expected primary-expression before \xe2\x80\x98<<\xe2\x80\x99 token
/mnt/disk1/home/vamosax/smoketest/PR-10321/HPCC-Platform/plugins/unicodelib/unicodelib.cpp:85:5: error: expected primary-expression before \xe2\x80\x98<<\xe2\x80\x99 token
/mnt/disk1/home/vamosax/smoketest/PR-10321/HPCC-Platform/plugins/unicodelib/unicodelib.cpp:85:7: error: expected primary-expression before \xe2\x80\x98<\xe2\x80\x99 token
/mnt/disk1/home/vamosax/smoketest/PR-10321/HPCC-Platform/plugins/unicodelib/unicodelib.cpp:85:9: error: \xe2\x80\x98HEAD\xe2\x80\x99 was not declared in this scope
/mnt/disk1/home/vamosax/smoketest/PR-10321/HPCC-Platform/plugins/unicodelib/unicodelib.cpp:726:1: error: expected unqualified-id before \xe2\x80\x98<<\xe2\x80\x99 token
/mnt/disk1/home/vamosax/smoketest/PR-10321/HPCC-Platform/plugins/unicodelib/unicodelib.cpp:1522:1: error: expected \xe2\x80\x98}\xe2\x80\x99 at end of input
make[2]: *** [plugins/unicodelib/CMakeFiles/unicodelib.dir/unicodelib.cpp.o] Error 1
make[1]: *** [plugins/unicodelib/CMakeFiles/unicodelib.dir/all] Error 2
[  0%] Building CXX object aws-cpp-sdk-core/CMakeFiles/aws-cpp-sdk-core.dir/source/client/CoreErrors.cpp.o
[100%] Building CXX object aws-cpp-sdk-sqs/CMakeFiles/aws-cpp-sdk-sqs.dir/source/SQSErrors.cpp.o
make: *** [all] Error 2
Get ECLWatch build result

HPCCSmoketest commented:
Automated Smoketest: :x:
Sha: 0c19b22d46b6c4d99af9de3835617f9b0be48806
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 75 | 45.089 seconds |


Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |91 |91 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |744 |743 |1 |
|test (thor) |639 |638 |1 |
|test (roxie) |772 |771 |1 |


[Errors:](https://gist.github.com/79dede51cbc971a8adab07da68edf710)
- test
  - hthor
    - 675\. [teststdlibrary](https://gist.github.com/HPCCSmoketest/79dede51cbc971a8adab07da68edf710#file-teststdlibrary-no-log), 
  - thor
    - 587\. [teststdlibrary](https://gist.github.com/HPCCSmoketest/79dede51cbc971a8adab07da68edf710#file-teststdlibrary-no-log), 
  - roxie
    - 708\. [teststdlibrary](https://gist.github.com/HPCCSmoketest/79dede51cbc971a8adab07da68edf710#file-teststdlibrary-no-log), 

HPCC Stop: OK
HPCC Uninstall: OK

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 283ac6405735044bd038b8705a4ffebfafa20981
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 75 | 44.318 seconds |


Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |91 |91 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |744 |744 |0 |
|test (thor) |639 |639 |0 |
|test (roxie) |772 |772 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

Commit:
HPCC-18041 Unicode Implementation for ExcludeLastWord

- Adds testcases to ecllibarry/teststd/uni/TestExcludeLastWord.ecl
- Adds documentation to ecllibrary/std/Uni.ecl
- Adds code to plugins/unicodelib.cpp

Signed-off-by: David Skaff <david.skaff@lexisnexisrisk.com>

dskaff commented:
@johnholt please code review.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 88980399d9751a72b2b6f6fa75f20e284eac738b
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 75 | 42.3 seconds |


Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |91 |91 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |744 |744 |0 |
|test (thor) |639 |639 |0 |
|test (roxie) |772 |772 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

ghalliday commented on plugins/unicodelib/unicodelib.cpp:
```diff
@@ -719,6 +720,42 @@ unsigned unicodeEditDistanceV4(UnicodeString & left, UnicodeString & right, unsi
+            {
+                ++wordidx;
+            }
+            if (bi.previous() == bi.first() && wordidx == 1)
```
This isn't safe.  Both bi.first() and bi.previous() modify the current iterator position.  It is undefined which is called first.

I think you should be comparing against 0.

ghalliday replied:
It isn't clear what this code is here for - there should be a comment to describe why

ghalliday replied:
Actually documentation says:  
The character index of the preceding text boundary, or UBRK_DONE if all text boundaries have been returned.

I'm not sure what you are testing here.

dskaff replied:
Normally, the function only removes the trailing white spaces of the last word, but when the source string only contains one word, the leading and trailing white spaces are removed (according to the string test cases).
The condition sets the iterator at the position previous to the beginning of the last word. If the string is only one word, then that position should be equal to bi.first().

dskaff replied:
I'll change bi.first() to 0

dskaff replied:
I don't believe that it will return UBRK_DONE unless you are using bi.preceding() while on the first boundary.

ghalliday commented on plugins/unicodelib/unicodelib.cpp:
```diff
@@ -719,6 +720,42 @@ unsigned unicodeEditDistanceV4(UnicodeString & left, UnicodeString & right, unsi
     return da[mask(leftLen-1)][rightLen-1];
 }
 
+void excludeLastWord(RuleBasedBreakIterator& bi, UnicodeString & source)
```
style: Not keen on a parameter that is modified being called source (see other PRs for comments)

ghalliday reviewed (changes requested):
@dskaff a good first stab, but I have some concerns about some of the code.

ghalliday commented:
Also checkboxes are not completed on the PR.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 0858c91c46139dbca7c065f7866f9b45d9374378
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 75 | 42.965 seconds |


Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |91 |91 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |744 |744 |0 |
|test (thor) |639 |639 |0 |
|test (roxie) |772 |772 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

dskaff commented:
@ghalliday please code review

ghalliday commented on plugins/unicodelib/unicodelib.cpp:
```diff
@@ -719,6 +720,39 @@ unsigned unicodeEditDistanceV4(UnicodeString & left, UnicodeString & right, unsi
+            //   not start a word or number.
+            ++wordidx;
+            wordBeginning = bi.previous();
+            if (bi.getRuleStatus() != UBRK_WORD_NONE)
```
This needs a comment explaining the logic - I can't follow it - or why bi.previous() would be called twice before checking it is 0.

dskaff commented:
@ghalliday I've commented the code

HPCCSmoketest commented:
Automated Smoketest: :x:
Sha: 9a4ac55b4f423575383931c6b28dc19a1a3d0325
Build: failed
Number of Error(s): 14
/mnt/disk1/home/vamosax/smoketest/PR-10321/HPCC-Platform/plugins/unicodelib/unicodelib.hpp:105:1: error: expected unqualified-id before \xe2\x80\x98<<\xe2\x80\x99 token
/mnt/disk1/home/vamosax/smoketest/PR-10321/HPCC-Platform/plugins/unicodelib/unicodelib.hpp:107:1: error: expected unqualified-id before \xe2\x80\x98==\xe2\x80\x99 token
/mnt/disk1/home/vamosax/smoketest/PR-10321/HPCC-Platform/plugins/unicodelib/unicodelib.hpp:109:1: error: expected unqualified-id before \xe2\x80\x98>>\xe2\x80\x99 token
/mnt/disk1/home/vamosax/smoketest/PR-10321/HPCC-Platform/plugins/unicodelib/unicodelib.cpp:86:3: error: expected primary-expression before \xe2\x80\x98<<\xe2\x80\x99 token
/mnt/disk1/home/vamosax/smoketest/PR-10321/HPCC-Platform/plugins/unicodelib/unicodelib.cpp:86:5: error: expected primary-expression before \xe2\x80\x98<<\xe2\x80\x99 token
/mnt/disk1/home/vamosax/smoketest/PR-10321/HPCC-Platform/plugins/unicodelib/unicodelib.cpp:86:7: error: expected primary-expression before \xe2\x80\x98<\xe2\x80\x99 token
/mnt/disk1/home/vamosax/smoketest/PR-10321/HPCC-Platform/plugins/unicodelib/unicodelib.cpp:86:9: error: \xe2\x80\x98HEAD\xe2\x80\x99 was not declared in this scope
/mnt/disk1/home/vamosax/smoketest/PR-10321/HPCC-Platform/plugins/unicodelib/unicodelib.cpp:727:1: error: expected unqualified-id before \xe2\x80\x98<<\xe2\x80\x99 token
/mnt/disk1/home/vamosax/smoketest/PR-10321/HPCC-Platform/plugins/unicodelib/unicodelib.cpp:1586:1: error: expected \xe2\x80\x98}\xe2\x80\x99 at end of input
make[2]: *** [plugins/unicodelib/CMakeFiles/unicodelib.dir/unicodelib.cpp.o] Error 1
make[1]: *** [plugins/unicodelib/CMakeFiles/unicodelib.dir/all] Error 2
[  0%] Building CXX object aws-cpp-sdk-core/CMakeFiles/aws-cpp-sdk-core.dir/source/client/CoreErrors.cpp.o
[100%] Building CXX object aws-cpp-sdk-sqs/CMakeFiles/aws-cpp-sdk-sqs.dir/source/SQSErrors.cpp.o
make: *** [all] Error 2
Get ECLWatch build result

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 43e3d7ea316c1c36f816cce2d9008f898131ad9e
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 75 | 46.405 seconds |


Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |91 |91 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |744 |744 |0 |
|test (thor) |639 |639 |0 |
|test (roxie) |772 |772 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
